core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,274 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A skeletal {@link Processor} implementation that decodes a stream of {@link HttpObject}s to N objects.
+ */
+@UnstableApi
+public abstract class HttpDeframer<T> extends DefaultStreamMessage<T> implements Processor<HttpObject, T> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<HttpDeframer, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(HttpDeframer.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> initializedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "initialized");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "subscribed");
+
+    private final ByteBufDeframerInput input;
+    private final ArrayDeque<T> outputQueue = new ArrayDeque<>();
+
+    private boolean sawLeadingHeaders;
+
+    @Nullable
+    private volatile EventExecutor eventLoop;
+    @Nullable
+    private volatile Subscription upstream;
+    private volatile boolean cancelled;
+    private volatile int initialized;
+    private volatile int subscribed;
+
+    /**
+     * Returns a new {@link HttpDeframer} with the specified {@link EventExecutor} and
+     * {@link ByteBufAllocator}.
+     */
+    protected HttpDeframer(ByteBufAllocator alloc) {
+        requireNonNull(alloc, "alloc");
+        input = new ByteBufDeframerInput(alloc);
+    }
+
+    /**
+     * Converts the specified {@link HttpData} to {@link ByteBuf}.
+     */
+    protected ByteBuf convertToByteBuf(HttpData data) {
+        return data.byteBuf();
+    }
```

Comment:
Question: Is this meant to be overridden?

Reply:
Before adding the `ByteBuf` to HttpDeframerInput, a Base64 encoded ByteBuf needs to be decoded.
https://github.com/line/armeria/blob/fb091d30f5ad3b189f185bc744911dc59c176634/grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageDeframer.java#L215-L222